Standards/scs-0214-v2-k8s-node-distribution.md

```diff
@@ -0,0 +1,130 @@
+[Kubernetes Nodes Anti Affinity][scs-0213-v1] as well as the Kubernetes documents about
+[High Availability][k8s-ha] and [Best practices for large clusters][k8s-large-clusters].
+
+SCS wishes to prefer distributed, highly-available systems due to their obvious advantages
+like fault-tolerance and data redundancy. But it also understands the costs and overhead
+for the providers associated with this effort, since the infrastructure needs to have
+hardware which will just be used to provide fail-over safety or duplication.
+
+The document [Best practices for large clusters][k8s-large-clusters] describes the concept of a failure zone.
+This term isn't defined any further, but can in this context be described as a number of
+physical (computing) machines in such a vicinity to each other (either through physical
+or logical interconnection in some way), that specific problems inside this zone would put
+all these machines at risk of failure/shutdown. It is therefore necessary for important
+data or services to not be present just on one failure zone.
+How such a failure zone should be defined is dependent on the risk model of the service/data
+and its owner as well as the capabilities of the provider. Zones could be set from things
+like single machines or racks up to whole datacenters or even regions, which could be
+coupled by things like electrical grids. They're therefore purely logical entities, which
+shouldn't be defined further in this document.
+
+## Decision
+
+This standard formulates the requirement for the distribution of Kubernetes nodes in order
+to provide a fault-tolerant and available Kubernetes cluster infrastructure.
+Since some providers only have small environments to work with and therefore couldn't
+comply with this standard, it will be treated as a RECOMMENDED standard, where providers
+can OPT OUT.
+
+If the standard is used by a provider, the following decisions are binding and valid:
+
+- The control plane nodes MUST be distributed over multiple physical machines. Kubernetes
+  provides best-practices on this topic, which are also RECOMMENDED by SCS.
+- At least one control plane instance MUST be run in each "failure zone", more are
+  RECOMMENDED in each "failure zone" to provide fault-tolerance for each zone.
+- Worker nodes are RECOMMENDED to be distributed over multiple zones. This policy makes
+  it OPTIONAL to provide a worker node in each "failure zone", meaning that worker nodes
+  can also be scaled vertically first before scaling horizontally.
+- Worker node distribution MUST be indicated to the user through some kind of labeling
+  in order to enable (anti)-affinity for workloads over "failure zones".
+- To provide metadata about the node distribution, which also enables testing of this standard,
+  providers MUST annotate their K8s nodes with the labels listed below. These labels MUST be kept
+  up to date with the current state of the deployment.
```

Comment:
As of now, we don't have any idea when this can be implemented with CAPO. @matofeder Do you have an update?
Unless the implemenation is in sight, we should demote this from a "MUST annotate" to a "SHOULD annotate". (The "MUST be kept up to date" can remain, so long as it's clear that it only applies if labels are being set at all.)

Reply:
@chess-knight as you have better visibility of this topic now, please provide an update here, thx

Reply:
The draft PR https://github.com/SovereignCloudStack/cluster-stacks/pull/53 with hostId labeling workaround was created by @matofeder and @SovereignCloudStack/vp06c will continue that work ASAP.
Regarding upstream, there is no progress AFAIK.

Reply:
I created an upstream issue https://github.com/kubernetes/cloud-provider-openstack/issues/2579 so let's wait for their opinion on that.

Reply:
Since this seems to be a low-hanging fruit in the opinion of the upstream guys, should I/we put the code in ourselves in order to speed the process up (https://github.com/kubernetes/cloud-provider-openstack/issues/2579) ?

Reply:
PR https://github.com/SovereignCloudStack/cluster-stacks/pull/53 is ready for review and it includes a workaround (creating server groups, instructing CAPO to use these server groups for nodes, labelling nodes with custom host-id label). https://github.com/kubernetes/cloud-provider-openstack/issues/2579 is about labelling part. The server group part is still missing in CAPO. I will try to comment in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1912 to speed it up.

Reply:
Also, I am unsure if it is low-hanging fruit, as I wrote in https://github.com/kubernetes/cloud-provider-openstack/issues/2579 - Live migration can be problematic because the label will not be updated. IMO, it needs a discussion.

---

And what about bare-metal clusters? In that case, we do not need this specific label right?